### PR TITLE
MBM-273 Fire offensive all tiers and other changes

### PR DESCRIPTION
**Pull request by MattKirchh**
- Fix tick count for `onUsingTick` not being set properly server side
- Fix tick count not being reset between spell activations
- Added smoke particles to alloy furnace
- Lots of code clean up

### CODE REVIEW

**Comment by MattKirchh** [comment]
Look into twilight forest and dragons breath pool for potential networking insight.

File: src/main/java/com/mineboundteam/minebound/entity/FireProjectile.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import net.minecraft.world.entity.Mob;
+import net.minecraft.world.entity.projectile.Fireball;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.BaseFireBlock;
+import net.minecraft.world.phys.BlockHitResult;
+import net.minecraft.world.phys.EntityHitResult;
+import net.minecraft.world.phys.HitResult;
+import org.jetbrains.annotations.NotNull;
+
+public class FireProjectile extends Fireball {
```
Comment (insertStringHere) [inline src/main/java/com/mineboundteam/minebound/entity/FireProjectile.java:20]: Is there a reason we need a custom class instead of just using Fireball?

Comment (MattKirchh) [reply src/main/java/com/mineboundteam/minebound/entity/FireProjectile.java:20]: Can't modify the damage, whether or not it ignites blocks, or the maximum distance of a `Fireball`.

Comment (insertStringHere) [reply src/main/java/com/mineboundteam/minebound/entity/FireProjectile.java:24]: Oh, interesting

File: src/main/java/com/mineboundteam/minebound/magic/network/MagicButtonSync.java [hunk 2]
```diff
@@ -19,87 +16,75 @@
...
                     }
                     case PRIMARY_PRESSED -> {
-                        if (chest.getItem() instanceof MyrialArmorItem && ((MyrialArmorItem) chest.getItem()).getTier().handSlots > 0) {
+                        if (chest.getItem() instanceof MyrialArmorItem item && item.getTier().handSlots > 0) {
                             UseSpellHelper.useSpell(player, PlayerSelectedSpellsProvider.PRIMARY_SPELL);
-                            
+
                             int handStates = MagicEvents.playerStates.getOrDefault(player.getId(), 0);
                             handStates |= MagicAnimationSync.ArmUsersMsg.PRIMARY;
-                            MagicEvents.playerStates.put(player.getId(), handStates);
```
Comment (insertStringHere) [inline src/main/java/com/mineboundteam/minebound/magic/network/MagicButtonSync.java:45]: Why did you remove the put here?

Comment (MattKirchh) [reply src/main/java/com/mineboundteam/minebound/magic/network/MagicButtonSync.java:45]: Oops, accident from when I was debugging the arm issue. Fixed.

**Review by solidwaterslayer — APPROVED** [review]

**Comment by insertStringHere** [comment]
Not so much a change request as a question: the player can be hurt by their own fire. Do we want to add a condition to make them immune here? Or should we just add fire immunity to fire utility and encourage people to equip that if they're going to play dangerously?

**Review by insertStringHere — DISMISSED** [review]

File: src/main/java/com/mineboundteam/minebound/magic/OffensiveSpells/FireOffensiveSpell.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    @Override
+    public void onUsingTick(ItemStack stack, Level level, Player player, int tickCount) {
+        float yRot = player.getYRot();
+        float xRot = player.getXRot();
+        double x = (0 - Math.sin(yRot * Math.PI / 180f));
+        double y = (0 - Math.sin(xRot * Math.PI / 180f));
+        double z = (Math.cos(yRot * Math.PI / 180f));
+        for (int i = 0; i <= 360; i += 30) {
+            level.addFreshEntity(new FireProjectile(level,
+                    player.getX(),
```
Comment (insertStringHere) [inline src/main/java/com/mineboundteam/minebound/magic/OffensiveSpells/FireOffensiveSpell.java:70]: Do you think it would be relatively easy to shift this left or right depending on the using hand? So the origin looks like it's from where the player is casting it?

Comment (MattKirchh) [reply src/main/java/com/mineboundteam/minebound/magic/OffensiveSpells/FireOffensiveSpell.java:70]: I don't know how easy it will be, I'd have to take a shot at it to see which I can do.

**Comment by MattKirchh** [comment]
> Not so much a change request as a question: the player can be hurt by their own fire. Do we want to add a condition to make them immune here? Or should we just add fire immunity to fire utility and encourage people to equip that if they're going to play dangerously?

I can just add a check to `onHitEntity` in `FireProjectile`.

**Comment by MattKirchh** [comment]
> > Not so much a change request as a question: the player can be hurt by their own fire. Do we want to add a condition to make them immune here? Or should we just add fire immunity to fire utility and encourage people to equip that if they're going to play dangerously?
> 
> I can just add a check to `onHitEntity` in `FireProjectile`.

Don't actually need to add a check, just needed to set the `owner` of the projectile. The logic already existed for preventing the player from being hit by their own projectile.

File: src/main/java/com/mineboundteam/minebound/magic/OffensiveSpells/FireOffensiveSpell.java (hunk 3 above)
Comment (insertStringHere) [reply src/main/java/com/mineboundteam/minebound/magic/OffensiveSpells/FireOffensiveSpell.java:70]: you could probably hardcode the values; Do you want me to merge Light Utility so the animation works in singleplayer? You'll have to refactor the classes with your UsingTick changes

Comment (MattKirchh) [reply src/main/java/com/mineboundteam/minebound/magic/OffensiveSpells/FireOffensiveSpell.java:70]: Yeah, you can merge Light Utility.

**Review by solidwaterslayer — APPROVED** [review]

File: src/main/java/com/mineboundteam/minebound/magic/OffensiveSpells/FireOffensiveSpell.java (hunk 3 above)
Comment (MattKirchh) [reply src/main/java/com/mineboundteam/minebound/magic/OffensiveSpells/FireOffensiveSpell.java:70]: Just tried `ItemStack.isSameItemSameTags(getSelectedSpell(player, PlayerSelectedSpellsProvider.PRIMARY_SPELL), stack)` which has the opposite problem, it's always `true` even if the secondary spell is the one currently being used.

Comment (MattKirchh) [reply src/main/java/com/mineboundteam/minebound/magic/OffensiveSpells/FireOffensiveSpell.java:70]: Actually, I have an idea, I'll post an update once I've tried it.

Comment (MattKirchh) [reply src/main/java/com/mineboundteam/minebound/magic/OffensiveSpells/FireOffensiveSpell.java:70]: Alright, it worked. Required updating the parameters for all the using functions in `ActiveSpellItem`, but it works like a charm.

**Review by insertStringHere — APPROVED** [review]
